### PR TITLE
Refactor get topic names and types

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#361

### CODE REVIEW

File: rmw_connext_shared_cpp/src/shared_functions.cpp [hunk 1]
```diff
@@ -1053,32 +1296,38 @@ get_node_names(const char * implementation_identifier,
...
-    char * name = pbtd.participant_name.name;
+    const char * name = pbtd.participant_name.name;
     if (!name || dds_ret != DDS_RETCODE_OK) {
-      name = const_cast<char *>("(no name)");
+      name = "(no name)";
+    }
+    node_names->data[i] = rcutils_strdup(name, allocator);
+    if (!node_names->data[i]) {
+      RMW_SET_ERROR_MSG("could not allocate memory for node name")
+      return RMW_RET_BAD_ALLOC;
```
Comment (dirk-thomas) [inline rmw_connext_shared_cpp/src/shared_functions.cpp]: Needs to deallocate previously allocated `data[i]`. Same as for `rmw_fastrtps_cpp`.

Comment (wjwwood) [inline rmw_connext_shared_cpp/src/shared_functions.cpp]: Yup: f739056

**Comment by mikaelarguedas** [comment]
@wjwwood I noticed a few jobs with uninitialized warnings that seem related to this PR: 
http://ci.ros2.org/view/nightly/job/nightly_linux_release/478/warnings22Result/
https://github.com/ros2/rmw_connext/pull/231/files#diff-9bba412d4597100825da2b2307403d08R1092

**Comment by wjwwood** [comment]
Hmm, I fixed that warning in Fast-RTPS:

https://github.com/ros2/rmw_fastrtps/commit/027139a2103af9ae0a1e2e05d00258edfb627688#diff-8225b098f1ec4debcdcee03d65ae95d5R2906

I guess I missed this one. I'll make a pr to fix it.

Not sure how that got passed CI...

**Pull request merged**